CPP-5803 S7035 Add exception for casts to bool (#4394)
SonarSource/rspec: rules/S7035/cfamily/rule.adoc

=== Exceptions

The result of casting any integer type to `bool` does not depend on the specific integer type.

> For the same reason casting enumeration to `bool` does not depend on the underlying type of enumerator,
> and this rule does not raise issues for such cast.

Reviewer: ```suggestion
For the same reason, casting an enumeration to `bool` does not depend on the underlying type of enumerator, so this rule does not raise.
```

[source,cpp]
----
enum class Enum {A, B, C};
void foo(Enum e) {
  auto i = static_cast<bool>(e); // Compliant by exception, the intent is obvious
  ...
}
----

Unscoped enums with no underlying type specified have an implementation-defined implicit underlying type. 

> Calling `std::to_underlying` on them wouldn't make the code more portable so this rule doesn't raise on them.

Reviewer: ```suggestion
Calling `std::to_underlying` on them would not make the code more portable so this rule does not raise on them.
```
